File: cli/internal/connectapi/notification_preferences.go
package connectapi

import (
	"context"
	"strings"

	"connectrpc.com/connect"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
	corev1 "hmans.de/chatto/internal/pb/chatto/core/v1"
)

type notificationPreferencesService struct {
	api *API
}

func (s *notificationPreferencesService) GetServerNotificationPreference(ctx context.Context, _ *connect.Request[apiv1.GetServerNotificationPreferenceRequest]) (*connect.Response[apiv1.GetServerNotificationPreferenceResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	level, err := s.api.core.GetSpaceNotificationLevel(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}
	effectiveLevel := level
	if effectiveLevel == corev1.NotificationLevel_NOTIFICATION_LEVEL_UNSPECIFIED {
		effectiveLevel = corev1.NotificationLevel_NOTIFICATION_LEVEL_NORMAL
	}
	return connect.NewResponse(&apiv1.GetServerNotificationPreferenceResponse{
		Level:          coreNotificationLevelToAPI(level),
		EffectiveLevel: coreNotificationLevelToAPI(effectiveLevel),
	}), nil
}

func (s *notificationPreferencesService) UpdateServerNotificationPreference(ctx context.Context, req *connect.Request[apiv1.UpdateServerNotificationPreferenceRequest]) (*connect.Response[apiv1.UpdateServerNotificationPreferenceResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	level, err := apiNotificationLevelToCore(req.Msg.Level)
	if err != nil {
		return nil, err
	}

	if err := s.api.core.SetSpaceNotificationLevel(ctx, caller.UserID, level); err != nil {
		return nil, connectError(err)
	}
	effectiveLevel := level
	if effectiveLevel == corev1.NotificationLevel_NOTIFICATION_LEVEL_UNSPECIFIED {
		effectiveLevel = corev1.NotificationLevel_NOTIFICATION_LEVEL_NORMAL
	}
	return connect.NewResponse(&apiv1.UpdateServerNotificationPreferenceResponse{
		Level:          coreNotificationLevelToAPI(level),
		EffectiveLevel: coreNotificationLevelToAPI(effectiveLevel),
	}), nil
}

func (s *notificationPreferencesService) GetRoomNotificationPreference(ctx context.Context, req *connect.Request[apiv1.GetRoomNotificationPreferenceRequest]) (*connect.Response[apiv1.GetRoomNotificationPreferenceResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	if strings.TrimSpace(req.Msg.RoomId) == "" {
		return nil, invalidArgument("room_id is required")
	}

	pref, err := s.api.core.NotificationPreferences().GetRoomNotificationPreference(ctx, caller.UserID, req.Msg.RoomId)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.GetRoomNotificationPreferenceResponse{
		Level:          coreNotificationLevelToAPI(pref.Level),
		EffectiveLevel: coreNotificationLevelToAPI(pref.EffectiveLevel),
	}), nil
}

func (s *notificationPreferencesService) UpdateRoomNotificationPreference(ctx context.Context, req *connect.Request[apiv1.UpdateRoomNotificationPreferenceRequest]) (*connect.Response[apiv1.UpdateRoomNotificationPreferenceResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	if strings.TrimSpace(req.Msg.RoomId) == "" {
		return nil, invalidArgument("room_id is required")
	}
	level, err := apiNotificationLevelToCore(req.Msg.Level)
	if err != nil {
		return nil, err
	}

	pref, err := s.api.core.NotificationPreferences().SetRoomNotificationLevel(ctx, caller.UserID, req.Msg.RoomId, level)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.UpdateRoomNotificationPreferenceResponse{
		Level:          coreNotificationLevelToAPI(pref.Level),
		EffectiveLevel: coreNotificationLevelToAPI(pref.EffectiveLevel),
	}), nil
}

func apiNotificationLevelToCore(level apiv1.NotificationLevel) (corev1.NotificationLevel, error) {
	switch level {
	case apiv1.NotificationLevel_NOTIFICATION_LEVEL_DEFAULT:
		return corev1.NotificationLevel_NOTIFICATION_LEVEL_UNSPECIFIED, nil
	case apiv1.NotificationLevel_NOTIFICATION_LEVEL_MUTED:
		return corev1.NotificationLevel_NOTIFICATION_LEVEL_MUTED, nil
	case apiv1.NotificationLevel_NOTIFICATION_LEVEL_NORMAL:
		return corev1.NotificationLevel_NOTIFICATION_LEVEL_NORMAL, nil
	case apiv1.NotificationLevel_NOTIFICATION_LEVEL_ALL_MESSAGES:
		return corev1.NotificationLevel_NOTIFICATION_LEVEL_ALL_MESSAGES, nil
	default:
		return corev1.NotificationLevel_NOTIFICATION_LEVEL_UNSPECIFIED, invalidArgument("notification level must be DEFAULT, MUTED, NORMAL, or ALL_MESSAGES")
	}
}

func coreNotificationLevelToAPI(level corev1.NotificationLevel) apiv1.NotificationLevel {
	switch level {
	case corev1.NotificationLevel_NOTIFICATION_LEVEL_MUTED:
		return apiv1.NotificationLevel_NOTIFICATION_LEVEL_MUTED
	case corev1.NotificationLevel_NOTIFICATION_LEVEL_NORMAL:
		return apiv1.NotificationLevel_NOTIFICATION_LEVEL_NORMAL
	case corev1.NotificationLevel_NOTIFICATION_LEVEL_ALL_MESSAGES:
		return apiv1.NotificationLevel_NOTIFICATION_LEVEL_ALL_MESSAGES
	default:
		return apiv1.NotificationLevel_NOTIFICATION_LEVEL_DEFAULT
	}
}

File: apps/frontend/src/lib/api-client/notificationPreferences.ts
import {
  authHeaders,
  createChattoClient,
  handleAuthError,
  type ConnectAPIConfig as BaseConnectAPIConfig,
} from "./connect.js";
import { NotificationPreferencesService } from "@chatto/api-types/api/v1/notification_preferences_connect";
import { NotificationLevel } from "@chatto/api-types/api/v1/notification_preferences_pb";

export type ConnectAPIConfig = BaseConnectAPIConfig & {
  serverId: string;
};

export type NotificationPreference = {
  level: NotificationLevel;
  effectiveLevel: NotificationLevel;
};

export async function getServerNotificationPreference(
  config: ConnectAPIConfig,
): Promise<NotificationPreference> {
  const client = createNotificationPreferencesClient(config);
  let response;
  try {
    response = await client.getServerNotificationPreference(
      {},
      {
        headers: connectHeaders(config),
      },
    );
  } catch (err) {
    handleAuthError(config, err);
  }
  return {
    level: response.level,
    effectiveLevel: response.effectiveLevel,
  };
}

export async function updateServerNotificationPreference(
  config: ConnectAPIConfig,
  level: NotificationLevel,
): Promise<NotificationPreference> {
  const client = createNotificationPreferencesClient(config);
  let response;
  try {
    response = await client.updateServerNotificationPreference(
      { level },
      {
        headers: connectHeaders(config),
      },
    );
  } catch (err) {
    handleAuthError(config, err);
  }
  return {
    level: response.level,
    effectiveLevel: response.effectiveLevel,
  };
}

export async function updateRoomNotificationPreference(
  config: ConnectAPIConfig,
  roomId: string,
  level: NotificationLevel,
): Promise<NotificationPreference> {
  const client = createNotificationPreferencesClient(config);
  let response;
  try {
    response = await client.updateRoomNotificationPreference(
      {
        roomId,
        level,
      },
      {
        headers: connectHeaders(config),
      },
    );
  } catch (err) {
    handleAuthError(config, err);
  }
  return {
    level: response.level,
    effectiveLevel: response.effectiveLevel,
  };
}

function createNotificationPreferencesClient(config: ConnectAPIConfig) {
  return createChattoClient(NotificationPreferencesService, config);
}

function connectHeaders(config: ConnectAPIConfig) {
  return authHeaders(config);
}
